pdq/cpp/index/flat.h

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// ================================================================
+
+#ifndef FLAT_H
+#define FLAT_H
+
+#include <array>
+#include <pdq/cpp/common/pdqhashtypes.h>
+
+#ifdef __AVX512VPOPCNTDQ__
+#include <immintrin.h>
+#endif
+
+#include <stdint.h>
+#include <type_traits>
+#include <vector>
+namespace facebook {
+namespace pdq {
+namespace index {
+
+/**
+ * @brief Flat index for PDQ matching based on amortized linear scan.
+ *
+ * The implementation is similar to FAISS IndexBinaryFlat but with PDQ-specific
+ * memory layout optimizations, see issue #1810 for more details.
+ *
+ * No index building is required using this index, it simply needs a packed
+ * array of the database hashes fed through the matching function.
+ *
+ * Metadata can be tracked using parallel arrays.
+ *
+ * Live mutations can be done by partitioning the database into multiple
+ * subsets, then use atomic variables and other synchronization primitives to
+ * ensure thread-safety.
+ *
+ * This index is intended for use when:
+ *   - The query can be batched by 8 hashes at a time (the most common reason
+ *     is to search all 8 dihedral variants of an image), AND
+ *   - One of the following is true:
+ *     - Minimal memory usage is desired, OR
+ *     - The threshold varies widely (for example per-user risk-based
+ * thresholds) at runtime making MIH too inflexible / space-inefficient, OR
+ *     - Timing correlation between the content of the index and the query is
+ *       unacceptable (use test() instead), OR
+ *     - The runtime environment has a modern (2020+) AVX512 CPU with
+ *       AVX512VPOPCNTDQ support, which can significantly outperform 8 MIH
+ *       queries with much less pressure on microarchitectural resources,
+ *       testing shows ~3x~6x improvement with an index of 10 million hashes
+ *       depending on whether matches are found.
+ */
+class Flat {
+#ifdef __AVX512VPOPCNTDQ__
+  __m512i _packedNeedles[4];
+#else
+  std::array<facebook::pdq::hashing::Hash256, 8> _needles;
+#endif
+
+ public:
+#ifdef __AVX512VPOPCNTDQ__
+  static const bool SIMD_ACCELERATED = true;
+#else
+  static const bool SIMD_ACCELERATED = false;
+#endif
+
+#ifdef __AVX512VPOPCNTDQ__
+  Flat(const std::array<facebook::pdq::hashing::Hash256, 8>& needles) {
+    alignas(64) uint64_t tmp[8];
+
+    const __m512i ones = _mm512_set1_epi64(~0ULL);
+
+    // Split 256-bit hash into 4x64-bit words, then place each hash vertically
+    // in 64-bit lanes. Finally complement all registers.
+
+    for (size_t regi = 0; regi < 4; regi++) {
+      for (size_t needlei = 0; needlei < 8; needlei++) {
+        tmp[needlei] =
+            reinterpret_cast<const uint64_t*>(needles[needlei].w)[regi];
+      }
+      _packedNeedles[regi] = _mm512_xor_epi64(ones, _mm512_load_si512(tmp));
+    }
+  }
+#else
+  Flat(const std::array<facebook::pdq::hashing::Hash256, 8>& needles)
+      : _needles(needles) {}
+#endif
+
+  ~Flat() {}
```

Comment:
```suggestion
   ~Flat() = default;
```

`~Flat() {}` --> ``,

See https://github.com/facebook/ThreatExchange/pull/1817/files#r2065413659